stackslib/src/net/api/tests/getclaritymetadata.rs

```diff
@@ -0,0 +1,273 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+
+use clarity::vm::database::{ClaritySerializable, DataMapMetadata, DataVariableMetadata};
+use clarity::vm::types::{QualifiedContractIdentifier, StacksAddressExtensions, TypeSignature};
+use clarity::vm::{ClarityName, ContractName};
+use serde_json::json;
+use stacks_common::codec::StacksMessageCodec;
+use stacks_common::types::chainstate::StacksAddress;
+use stacks_common::types::net::PeerHost;
+use stacks_common::types::Address;
+
+use super::test_rpc;
+use crate::net::api::*;
+use crate::net::connection::ConnectionOptions;
+use crate::net::http::Error as HttpError;
+use crate::net::httpcore::{
+    HttpPreambleExtensions, HttpRequestContentsExtensions, RPCRequestHandler, StacksHttp,
+    StacksHttpRequest,
+};
+use crate::net::{Error as NetError, ProtocolFamily, TipRequest};
+
+#[test]
+fn test_try_parse_request() {
+    let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 33333);
+    let mut http = StacksHttp::new(addr.clone(), &ConnectionOptions::default());
+
+    let request = StacksHttpRequest::new_getclaritymetadata(
+        addr.into(),
+        StacksAddress::from_string("ST2DS4MSWSGJ3W9FBC6BVT0Y92S345HY8N3T6AV7R").unwrap(),
+        "hello-world".try_into().unwrap(),
+        "vm-metadata::9::contract-size".to_string(),
+        TipRequest::SpecificTip(StacksBlockId([0x22; 32])),
+    );
+    assert_eq!(
+        request.contents().tip_request(),
+        TipRequest::SpecificTip(StacksBlockId([0x22; 32]))
+    );
+    let bytes = request.try_serialize().unwrap();
+
+    let (parsed_preamble, offset) = http.read_preamble(&bytes).unwrap();
+    let mut handler = getclaritymetadata::RPCGetClarityMetadataRequestHandler::new();
+    let mut parsed_request = http
+        .handle_try_parse_request(
+            &mut handler,
+            &parsed_preamble.expect_request(),
+            &bytes[offset..],
+        )
+        .unwrap();
+
+    // parsed request consumes headers that would not be in a constructed request
+    parsed_request.clear_headers();
+    let (preamble, contents) = parsed_request.destruct();
+
+    // consumed path args
+    assert_eq!(
+        handler.clarity_metadata_key,
+        Some("vm-metadata::9::contract-size".to_string())
+    );
+    assert_eq!(
+        handler.contract_identifier,
+        Some(
+            QualifiedContractIdentifier::parse(
+                "ST2DS4MSWSGJ3W9FBC6BVT0Y92S345HY8N3T6AV7R.hello-world"
+            )
+            .unwrap()
+        )
+    );
+
+    assert_eq!(&preamble, request.preamble());
+
+    handler.restart();
+    assert!(handler.clarity_metadata_key.is_none());
```

Comment:
Can you expand this test to cover error cases?  Specifically, can you verify that the user gets back HTTP 400 for invalid metadata key?

Reply:
There are two test case for that below with `vm-metadata::2::contract-size` and `vm-metadata::9::contract-invalid-key`